orchestrator/lambdas/layers/heimdall_repos/heimdall_repos/github_utils.py

```diff
@@ -60,12 +61,16 @@ def _query_github_api(self, query: str) -> str:
             json={"query": query},
             timeout=DEFAULT_API_TIMEOUT,
         )
-        if response.status_code != 200:
-            error_response = self._analyze_error_response(response)
+
+        response_text = self._get_response_text(response)
+
+        # cannot depend on status code alone because rate limit errors have a 200 code
+        if response.status_code != 200 or self._check_for_errors_in_response(response_text):
+            error_response = self._analyze_error_response(response, response_text)
```

Comment:
Could we use the `/rate_limit` API endpoint to see definitively if the limit has been exceeded? https://docs.github.com/en/rest/rate-limit?apiVersion=2022-11-28#get-rate-limit-status-for-the-authenticated-user. Or, alternatively, check this API before making a request to ensure that there is capacity within the limit to be successful?

Reply:
Does the GitHub GraphQL API return the `X-RateLimit-Remaining`, `X-RateLimit-Reset`, and `Retry-After` response headers to indicate status related to the rate limit?

Reply:
Yes - all of these would work for checking rate limit status specifically. I'm thinking it would be an improvement to add a call to `/rate_limit` to `_is_message_rate_abuse()` (or create a separate function to check for this e.g. `_check_rate_limit_status()`).

However, I think there is still value in this generic implementation of `_check_for_errors_in_response()`, because we'd want it to catch any sort of error that GitHub's API sends back with a non-error status code (rate limiting is just the one I noticed when testing, but that's not to say it's the only one).

Reply:
Also, I can update the comment to something more general like-
`# cannot depend on status code alone because some errors have a 200 code (e.g. rate limit)`

Reply:
@pizen implemented a check of `X-RateLimit-Remaining` in `_analyze_error_response()` while retaining the generic check as well.